### DISCUSSION TITLE
Do I need one subject per event type, or one per domain?

### DISCUSSION THREAD

**Discussion by robertmain** [ReactiveX/rxjs#7022] [Q&A]
I'm not sure how to use `Subject` correctly. I'm building an application that uses (for example) a grocery list. I'm not sure if I should have different subjects for each type of action that may be performed on the grocery list like this:

```ts
const removeItem = new Subject<number>();
const addItem = new Subject<GroceryItem>();

removeItem.next(5);
addItem({ name: 'Eggs' });
```

or if I should instead have a `Subject` for the grocery list, and then somehow add a `type` property to my subject data to indicate the operation being performed like so:

```ts
type GroceryListAction = {
  type: 'ADD',
  data: { name: string }
} | {
  type: 'REMOVE'
  data: { id: number }
} ;


const groceryList = new Subject<GroceryListAction>();

groceryList.next({
  type: 'REMOVE',
  data: { id: 5 }
});
groceryList.next({
  type: 'ADD',
  data: { name: 'Eggs' }
});
```

**Comment by voliva** [comment]
They are equivalent. It's easy to build either of them from the other if you need to (merge + map for one side, filter + map for the other).

I personally use the first approach. Usually it has less boilerplate, and it lets me easily use those streams separately.
